record/record_test.go

```diff
@@ -312,3 +313,137 @@ func TestStaleWriter(t *testing.T) {
 		t.Fatalf("stale write #1: unexpected error: %v", err)
 	}
 }
+
+func TestBasicReadRecovery(t *testing.T) {
+	buf := new(bytes.Buffer)
+	w := NewWriter(buf)
+
+	for i := 0; i < 3; i++ {
+		if wRec, err := w.Next(); err != nil {
+			t.Fatal(err)
+		} else {
+			if _, err = wRec.Write([]byte(strings.Repeat(string(rune(97+i)), blockSize-headerSize))); err != nil {
+				t.Fatal(err)
+			}
+		}
+	}
+	w.Close() // Finalize the last record.
+
+	rawBufSlice := buf.Bytes()
+
+	// Corrupt the checksum of the second record in our file.
+	rawBufSlice[blockSize] = 0xEF
+	rawBufSlice[blockSize+1] = 0xBE
+	rawBufSlice[blockSize+2] = 0xAD
+	rawBufSlice[blockSize+3] = 0xDE
+
+	// The first record should be read/processed just fine.
+	underlyingReader := bytes.NewReader(rawBufSlice)
+	r := NewReader(underlyingReader)
+	r1, err := r.Next()
+	if err != nil {
+		t.Fatal(err)
+	}
+	r1Data, err := ioutil.ReadAll(r1)
+	if bytes.Compare(r1Data, []byte(strings.Repeat("a", blockSize-headerSize))) != 0 {
+		t.Fatal("Unexpected output in record 1's data.")
+	}
+
+	// This record should present problems since the checksum is wrong.
+	_, err = r.Next()
+	if err == nil {
+		t.Fatal(err)
+	}
+
+	if !strings.Contains(err.Error(), "checksum mismatch") {
+		t.Fatalf("Unexpected error returned: %s", err.Error())
+	}
+
+	// Now attempt to recover from the checksum error we intentionally caused.
+	err = r.Recover()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	currentOffset, err := underlyingReader.Seek(0, os.SEEK_CUR)
```

Comment:
It's not, but I think it's important to do so because the cursor should end up at the right place after calling Recover.  It's possible Recover is broken and ends up two blocks off, instead of one. This makes sure it goes to the right spot.  At least that's my thinking.